controller/snapshot_controller.go

```diff
@@ -0,0 +1,585 @@
+
+	ds                     *datastore.DataStore
+	cacheSyncs             []cache.InformerSynced
+	engineClientCollection engineapi.EngineClientCollection
+
+	monitor *SnapshotMonitor
+}
+
+type SnapshotCache struct {
+	logger logrus.FieldLogger
+
+	namespace string
+	ds        *datastore.DataStore
+
+	controllerID string
+
+	mutex         sync.RWMutex
+	snapshotStore map[string]*longhorn.Snapshot
+}
+
+type SnapshotMonitor struct {
+	logger logrus.FieldLogger
+
+	namespace string
+	ds        *datastore.DataStore
+
+	controllerID string
+
+	mutex         sync.RWMutex
+	snapshotStore map[string]*longhorn.Snapshot
+}
+
+func NewSnapshotController(
+	logger logrus.FieldLogger,
+	ds *datastore.DataStore,
+	scheme *runtime.Scheme,
+	kubeClient clientset.Interface,
+	namespace string,
+	controllerID string,
+	engineClientCollection engineapi.EngineClientCollection) *SnapshotController {
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(logrus.Infof)
+	// TODO: remove the wrapper when every clients have moved to use the clientset.
+	eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{
+		Interface: v1core.New(kubeClient.CoreV1().RESTClient()).Events(""),
+	})
+
+	sc := &SnapshotController{
+		baseController: newBaseController("longhorn-snapshot", logger),
+
+		namespace:              namespace,
+		controllerID:           controllerID,
+		kubeClient:             kubeClient,
+		eventRecorder:          eventBroadcaster.NewRecorder(scheme, v1.EventSource{Component: "longhorn-snapshot-controller"}),
+		ds:                     ds,
+		engineClientCollection: engineClientCollection,
+	}
+
+	sc.monitor = &SnapshotMonitor{
+		logger:        sc.logger.WithField("component", "snapshotMonitor"),
+		namespace:     namespace,
+		ds:            ds,
+		controllerID:  controllerID,
+		snapshotStore: make(map[string]*longhorn.Snapshot),
+	}
+
+	ds.SnapshotInformer.AddEventHandlerWithResyncPeriod(cache.ResourceEventHandlerFuncs{
+		AddFunc:    sc.enqueueSnapshot,
+		UpdateFunc: func(old, cur interface{}) { sc.enqueueSnapshot(cur) },
+		DeleteFunc: sc.enqueueSnapshot,
+	}, 0)
+	sc.cacheSyncs = append(sc.cacheSyncs, ds.SnapshotInformer.HasSynced)
+	ds.EngineInformer.AddEventHandlerWithResyncPeriod(cache.ResourceEventHandlerFuncs{
+		UpdateFunc: sc.enqueueEngineChange,
+	}, 0)
+
+	return sc
+}
+
+func (sc *SnapshotController) enqueueSnapshot(obj interface{}) {
+	key, err := controller.KeyFunc(obj)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %#v: %v", obj, err))
+		return
+	}
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return
+	}
+	if namespace != sc.namespace {
+		return
+	}
+
+	snap, err := sc.ds.GetSnapshot(name)
+	if err != nil {
+		sc.queue.Add(key)
+		return
+	}
+
+	vol, err := sc.ds.GetVolumeRO(snap.Spec.Volume)
+	if err != nil {
+		sc.queue.Add(key)
+		return
+	}
+
+	if vol.Status.OwnerID != sc.controllerID {
+		return
+	}
```

Comment:
NIT: Should we check these here?

Reply:
It helps to reduce the workqueue depth significantly. In my test from 20k down to 2k (10 worker nodes)

Reply:
update. The 20k queue was due to the worker queue cannot take out key fast enough. Agree that we should keep this enqueue path simple. PR is created at https://github.com/longhorn/longhorn-manager/pull/1266